### ISSUE TITLE
torch.compile breaks reproducibility

### ISSUE THREAD

**Issue by rahul-1996** [pytorch/pytorch#94855]
### 🐛 Describe the bug

Adding torch.compile does not ensure deterministic results after setting a seed (and ensuring all the steps here: https://pytorch.org/docs/stable/notes/randomness.html#:~:text=Reproducibility%20Completely%20reproducible%20results%20are%20not%20guaranteed%20across,and%20GPU%20executions%2C%20even%20when%20using%20identical%20seeds.). 

I've been stuck trying to debug why my results are non-deterministic across runs. Finally, removing torch.compile ensures that results across multiple runs are the same. This can be easily reproduced by having multiple runs of a model with torch.compile enabled. 

### Versions

PyTorch version: 2.0.0.dev20230213+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.15 (default, Nov 24 2022, 15:19:38)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-1085-azure-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe

Nvidia driver version: 510.73.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              1
Core(s) per socket:              48
Socket(s):                       2
NUMA node(s):                    4
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           1
Model name:                      AMD EPYC 7V13 64-Core Processor
Stepping:                        1
CPU MHz:                         2478.466
BogoMIPS:                        4890.88
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       3 MiB
L1i cache:                       3 MiB
L2 cache:                        48 MiB
L3 cache:                        384 MiB
NUMA node0 CPU(s):               0-23
NUMA node1 CPU(s):               24-47
NUMA node2 CPU(s):               48-71
NUMA node3 CPU(s):               72-95
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, STIBP disabled, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core invpcid_single vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr arat umip vaes vpclmulqdq rdpid

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==1.6.3
[pip3] pytorch-triton==2.0.0+0d7e753227
[pip3] torch==2.0.0.dev20230213+cu117
[pip3] torch-nebula==0.15.9
[pip3] torch-ort==1.13.1
[pip3] torchaudio==2.0.0.dev20230214+cu117
[pip3] torchmetrics==0.7.1
[pip3] torchvision==0.15.0.dev20230214+cu117
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2021.4.0                 pypi_0    pypi
[conda] mkl-include               2021.4.0                 pypi_0    pypi
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-lightning         1.6.3                    pypi_0    pypi
[conda] pytorch-triton            2.0.0+0d7e753227          pypi_0    pypi
[conda] torch                     2.0.0.dev20230213+cu117          pypi_0    pypi
[conda] torch-nebula              0.15.9                   pypi_0    pypi
[conda] torch-ort                 1.13.1                   pypi_0    pypi
[conda] torchaudio                2.0.0.dev20230214+cu117          pypi_0    pypi
[conda] torchmetrics              0.7.1                    pypi_0    pypi
[conda] torchvision               0.15.0.dev20230214+cu117          pypi_0    pypi

**Comment by ngimel** [comment]
Do you have some index_put or scatter_reduce operations in your workload?

**Comment by rahul-1996** [comment]
None, at least not explicitly doing those operations.

**Comment by msaroufim** [comment]
@rahul-1996 do you have a specific model you could share?

You should expect some degree of changes between a torch.compiled model vs an eager model because randomness is generated differently across eager and the inductor backend with triton

**Comment by rahul-1996** [comment]
@msaroufim I am training a BERT variant from huggingface: https://huggingface.co/bert-base-uncased. Maybe has something to do with the huggingface implementation.

**Comment by msaroufim** [comment]
Does setting this to True fix your issue? https://github.com/pytorch/pytorch/blob/master/torch/_inductor/config.py#L61

**Comment by ezyang** [comment]
Can you try again? We've made inductor a lot more deterministic in the meantime

**Comment by Hamlynn** [comment]
I've been stuck in the same problem, have you sovled it?

**Comment by Superhzf** [comment]
I have the same issue on PyTorch 2.0.1

**Comment by ezyang** [comment]
please try nightly, most of the improvements haven't made it to official release yet

**Comment by cisimon7** [comment]
I am still noticing this in my training script with PyTorch 2.1.1

**Comment by Oktai15** [comment]
@ngimel @msaroufim @ezyang I can share minimal code to reproduce mentioned problem.

You can try to run this function:
```python
# @torch.compile
def slice_speech_representation(
    speech_representation: torch.Tensor,
    speech_representation_lengths: torch.Tensor,
    slice_length: int,
):
    """
    Args:
        speech_representation: [batch_size, H, T]
        speech_representation_lengths: [batch_size, ]
        slice_length: int

    Returns:
        slice_speech_representation: [batch_size, H, slice_length]
        slice_indices: [batch_size, slice_length]
    """
    batch_size, h, _ = speech_representation.shape
    device = speech_representation.device

    # Generate random start indices for each sample in the batch
    max_start_indices = speech_representation_lengths - slice_length
    start_indices = (torch.rand(batch_size, device=device) * max_start_indices).long()

    # Create a range tensor for slicing and adjust with start indices
    slice_indices = (torch.arange(slice_length, device=device) + start_indices[:, None]).long()

    # Gather the slices for speech representation
    slice_speech_repr = torch.gather(
        speech_representation,
        2,
        slice_indices.unsqueeze(1).expand(-1, h, -1),
    )

    return slice_speech_repr, slice_indices

torch.manual_seed(0)
torch.cuda.manual_seed(0)

device = torch.device("cuda:0")
audio = torch.randn(2, 1, 1000).to(device)
lengths = torch.randint(200, 1000, (2,)).to(device)

torch.manual_seed(0)
torch.cuda.manual_seed(0)

slice_speech_repr, slice_indices = slice_speech_representation(audio, lengths, 100)
```

Please, try to run this code, check output, after that uncomment @torch.compile and you can use that output is different.

**Comment by Oktai15** [comment]
This flag returns reproducibility:
```python
from torch._inductor import config
config.fallback_random = True
```

**Comment by ezyang** [comment]
Putting this in triage review to ask inductor peeps if there are known nondeterminism in our inductor rng codegen.

**Comment by mattrmd** [comment]
From tests on my devices/models, it seems like `torch.compile` doesn't introduce reproducibility issues. I was able to get identical cached models on two different runs, as measured by `sha356sum` I'm running `torch==2.3.1` with Python 3.10.14. I have this at the top of my Python script, which I found on the [PyTorch](https://pytorch.org/docs/stable/notes/randomness.html) and [NVIDIA](https://docs.nvidia.com/cuda/cublas/index.html#results-reproducibility) websites:
```python
import os
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"

import torch
torch.backends.cudnn.benchmark = False
torch.use_deterministic_algorithms(True)
torch.random.manual_seed(0)
```

**Comment by eellison** [comment]
We don't have repros of this so un-assigning. The settings @mattrmd posted should help, and numerics with rng between inductor and eager will differ unless you set `config.fallback_random = True`.

**Comment by TomQunChao** [comment]
> We don't have repros of this so un-assigning. The settings [@mattrmd](https://github.com/mattrmd) posted should help, and numerics with rng between inductor and eager will differ unless you set `config.fallback_random = True`.

I ran into the same issue. After setting this flag, the a/a noise can be eliminated on H20, but it cannot be eliminated on H200.

Environment: cuda 12.4 / torch 2.6.0

I previously tried disabling TF32 or setting cudnn.deterministic=True, which also eliminated the a/a noise. So my guess about the role of this flag is that it more likely changes CUDA stream scheduling rather than the determinism of the kernel itself.
